crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -0,0 +1,325 @@
+            .eth()
+            .transaction(TransactionId::Hash(hash))
+            .await
+            .with_context(|| format!("get tx {hash:?}"))?
+            .with_context(|| format!("no tx {hash:?}"))?;
+        let tx_from = transaction
+            .from
+            .with_context(|| format!("no from {hash:?}"))?;
+        let tx_nonce: i64 = transaction
+            .nonce
+            .try_into()
+            .map_err(|err| anyhow!("{}", err))
+            .with_context(|| format!("convert nonce {hash:?}"))?;
+        let receipt = self
+            .web3
+            .eth()
+            .transaction_receipt(hash)
+            .await?
+            .with_context(|| format!("no receipt {hash:?}"))?;
+        let gas_used = receipt
+            .gas_used
+            .with_context(|| format!("no gas used {hash:?}"))?;
+        let effective_gas_price = receipt
+            .effective_gas_price
+            .with_context(|| format!("no effective gas price {hash:?}"))?;
+
+        let auction_id = self
+            .db
+            .get_auction_id(tx_from, tx_nonce)
+            .await?
+            .context(format!("no auction id for tx {hash:?}"))?;
+        let auction_external_prices =
+            self.db
+                .get_auction_prices(auction_id)
+                .await
+                .context(format!(
+                    "no external prices for auction id {auction_id:?} and tx {hash:?}"
+                ))?;
+        let orders = self
+            .db
+            .orders_for_tx(&hash)
+            .await?
+            .into_iter()
+            .map(Into::into)
+            .collect::<Vec<_>>();
+        let external_prices = ExternalPrices::try_from_auction_prices(
+            self.native_token,
+            auction_external_prices.clone(),
+        )?;
+
+        // surplus and fees calculation
+        let configuration = FeeConfiguration {
+            fee_objective_scaling_factor: self.fee_objective_scaling_factor.clone(),
+        };
+        let settlement = DecodedSettlement::new(&transaction.input.0)?;
+        let surplus = settlement.total_surplus(&external_prices);
+        let fee = settlement.total_fees(&external_prices, &orders, &configuration);
+
+        // reduce external prices in `auction_prices` table to only include used tokens
+        // this is done to reduce the amount of data we store in the database
+        let prices = settlement
+            .trades
+            .iter()
+            .filter_map(|trade| {
+                let buy_token = settlement
+                    .tokens
+                    .get(trade.buy_token_index.as_u64() as usize)
+                    .unwrap(); // TODO can I unwrap here? Is it guaranteed by the settlement contract?
+                let sell_token = settlement
+                    .tokens
+                    .get(trade.sell_token_index.as_u64() as usize)
+                    .unwrap(); // TODO can I unwrap here? Is it guaranteed by the settlement contract?
```

Comment:
Seems reasonable to me. I guess then you could fetch them by index directly instead of via "get"

Reply:
Removed unwrap.